TLS Client certificate support for Android and WearOS APP

mircoboschi opened the pull request:
## Summary
Add Client TLS support to Android and WearOS APP.
This allow to have a NGINX/Cloudflare reverse proxy in front of Home Assistant that accepts connections only from clients having a valid certificate.
Without certificate, Home Assistant is kept completely unreachable

## Screenshots
No UI changes (not yet, at least)

## Link to pull request in Documentation repository
Documentation: https://github.com/home-assistant/companion.home-assistant/pull/636

## Any other notes
Probably cert file upload should be changed.
Client cert also works in Wear OS, but is tricky to upload cert files.

homeassistant commented:
Hello @mircoboschi,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`e97b907b98023406775f5d9a7fbdbf32bae9e013`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2023&#x2F;commits&#x2F;e97b907b98023406775f5d9a7fbdbf32bae9e013) - This commit has something that looks like an email address (mirco.boschi@flashinlabs.com). Maybe try linking that to GitHub?.
- [`7204a58ae540bd4cfafedd9588eb3096acd9617d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2023&#x2F;commits&#x2F;7204a58ae540bd4cfafedd9588eb3096acd9617d) - This commit has something that looks like an email address (mirco.boschi@flashinlabs.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @mircoboschi, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;android#2023).

Once you do that we will be able to review and accept this pull request.

Thanks!

dshokouhi commented on app/src/main/java/io/homeassistant/companion/android/util/MTLSWebViewClient.kt:
```diff
@@ -0,0 +1,43 @@
+
+    private fun showMissingClientCertError(context: Context) {
+        AlertDialog.Builder(context)
+            .setTitle("TLS Client certificate not available")
+            .setMessage(
+                "Remote site asked for client Certificate\nPlace tls_client.key and tls_client.crt in android/data/io.homeassistant.companion.android/files and restart application"
```
We should move these strings to resources so they can be translated

dshokouhi reviewed (changes requested):
Just a small comment, once the actions run I will also test the debug APK in my own setup

dshokouhi reviewed (commented):
Tested the debug APK and all seems to be well on my end without a client cert.  Will wait for other reviews and testers to see if we need to adjust the current implementation or not.

tormentist commented:
# Mutual TLS testing (WIP)
## Setup:  
apk used: `app-full-debug.apk`
Phone model: Samsung Note 9
Android 10, unrooted
nginx reverse proxy in front of home-assistant instance.  

example nginx config:  
`/etc/nginx/conf.d/hass.conf`:
```
server {
  listen 443 ssl;
  server_name <redacted>;
  ssl_protocols       TLSv1.2 TLSv1.3;
  ssl_certificate     "/root/.acme.sh/<redacted>/fullchain.cer";
  ssl_certificate_key "/root/.acme.sh/<redacted>/<redacted>.key";
  ssl_client_certificate /etc/nginx/client_certs/ca.crt;
  ssl_verify_client optional;

  location / {
    if ($ssl_client_verify != SUCCESS) {
      return 444;
    }

    resolver <redacted>;
    proxy_pass http://<redacted>:8123;
    proxy_set_header Host $host;
    proxy_redirect http:// https://;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_buffering off;
    proxy_ssl_verify off;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  }
}
```

## 1. Regression testing:  
Initial testing shows that not using mutual TLS seems to work as expected.  
Using the app without a client/key present and connecting to a traditional non-mutual-TLS URL works as expected.  

## 2. Mutual TLS testing (WIP) :  
### 2.1  First test:  having tls_client.crt and tls_client.key in the files folder results in a crash when opening the app.
Logs:  
```
12-18 21:41:08.636 30888 30888 E AndroidRuntime: Process: io.homeassistant.companion.android.debug, PID: 30888
12-18 21:41:08.636 30888 30888 E AndroidRuntime: java.lang.RuntimeException: Unable to create application io.homeassistant.companion.android.HomeAssistantApplication: java.security.spec.InvalidKeySpecException: com.android.org.conscrypt.OpenSSLX509CertificateFactory$ParsingException: Error parsing private key
12-18 21:41:08.636 30888 30888 E AndroidRuntime:        at io.homeassistant.companion.android.common.data.MTLSHelper.tryImportKeys(MTLSHelper.kt:64)
12-18 21:41:08.636 30888 30888 E AndroidRuntime:        at io.homeassistant.companion.android.common.data.MTLSHelper.init(MTLSHelper.kt:73)
12-18 21:41:08.636 30888 30888 E AndroidRuntime:        at io.homeassistant.companion.android.HomeAssistantApplication.onCreate(HomeAssistantApplication.kt:40)
12-18 21:41:08.640  1283  3868 W ActivityManager: crash : io.homeassistant.companion.android.debug,0
```
Relevant part: `Error parsing private key`  
Turns out my key file was encrypted / had a passphrase
I re-exported the cert and key without a passphrase
Note, an encrypted key header looks like this:  
```
-----BEGIN RSA PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
DEK-Info: DES-EDE3-CBC,C209488FF9751CEA

4lflt+gfJW <redacted>
```
Decrypted header looks like this after using `openssl rsa -in [original.key] -out [new.key]`:  
```
-----BEGIN RSA PRIVATE KEY-----
MIIJJ <redacted>
```
### 2.2 Second Test: decrypted tls_client.key and tls_client.crt present in the files folder results in a crash when opening the app.
Logs:  
```
12-18 22:40:50.912 12543 12543 E AndroidRuntime: Process: io.homeassistant.companion.android.debug, PID: 12543
12-18 22:40:50.912 12543 12543 E AndroidRuntime: java.lang.RuntimeException: Unable to create application io.homeassistant.companion.android.HomeAssistantApplication: java.security.cert.CertificateException: com.android.org.conscrypt.OpenSSLX509CertificateFactory$ParsingException: com.android.org.conscrypt.OpenSSLX509CertificateFactory$ParsingException: java.lang.RuntimeException: error:0c0000be:ASN.1 encoding routines:OPENSSL_internal:WRONG_TAG
12-18 22:40:50.912 12543 12543 E AndroidRuntime:        at io.homeassistant.companion.android.common.data.MTLSHelper.tryImportKeys(MTLSHelper.kt:57)
12-18 22:40:50.912 12543 12543 E AndroidRuntime:        at io.homeassistant.companion.android.common.data.MTLSHelper.init(MTLSHelper.kt:73)
12-18 22:40:50.912 12543 12543 E AndroidRuntime:        at io.homeassistant.companion.android.HomeAssistantApplication.onCreate(HomeAssistantApplication.kt:40)
12-18 22:40:50.915  1283  2097 W ActivityManager: crash : io.homeassistant.companion.android.debug,0
```

Relevant log: `OpenSSLX509CertificateFactory$ParsingException: java.lang.RuntimeException: error:0c0000be:ASN.1 encoding routines:OPENSSL_internal:WRONG_TAG`

Pretty sure this is another cert/key format problem.
Turns out `OpenSSLX509CertificateFactory` expects DER-encoded data, not PEM-encoded

Converting PEM to DER:  
```
openssl x509 -outform der -in tls_client.crt -out tls_client.der
mv tls_client.der tls_client.crt
```

### 2.3 Third test: decrypted tls_client.key and DER-encoded tls_client.crt present in the files folder... Works !
So far so good, I'll keep using it and report back if I find anything,
nginx server logs look OK too  

Switching from WiFi to LTE is seamless in my current setup too.
Although, I haven't tested the app's internal vs external URL features as I'm using split-horizon DNS instead.
I have a single proxied hass URL that resolves to my nginx's LAN IP when on LAN and to my public IP when outside my LAN with nginx running mutual TLS (ssl_verify_client) regardless whether connection is coming from LAN or WAN


And most importantly...
# Thanks !!!

tormentist commented:
Just noticed `networking.md` mentions a PEM-encoded tls_client.pem as the cert whereas the dialog box text in `common/src/main/res/values/strings.xml` shows tls_client.crt
From initial testing tls_client.pem isn't used and tls_client.crt needs to be DER-encoded, not sure if it's specific to my phone or not.

Also, I had a Firebase error about push notifications when logging in, I'll dig into it

Again, **thanks**!

SkechyWolf commented:
> Also, I had a Firebase error about push notifications when logging in, I'll dig into it

Debug apks don't have the needed files for notifications that's why it gives a firebase error I believe

mircoboschi commented:
> Just noticed `networking.md` mentions a PEM-encoded tls_client.pem as the cert whereas the dialog box text in `common/src/main/res/values/strings.xml` shows tls_client.crt From initial testing tls_client.pem isn't used and tls_client.crt needs to be DER-encoded, not sure if it's specific to my phone or not.
> 
> Also, I had a Firebase error about push notifications when logging in, I'll dig into it
> 
> Again, **thanks**!

Thanks for pointing typo ; documentation fixed.
Also added a note about no-password requirements.

mircoboschi commented:
For anyone testing this PR:
After initial setup, cert files can be removed from application folder because they are imported in AndroidKeyStore.
We could also automatically delete them once imported; any advice on this?

mircoboschi commented:
@tormentist About PEM/DER format issue:
my cert/key are PEM encoded like this, can you check if this differs from your PEM?
```
-----BEGIN CERTIFICATE-----
Base64DataHere
-----END CERTIFICATE-----
```
Can you please try to convert your certificate to pem format again with the following command?
`openssl x509 -in old_pem_cert.crt -out tls_client.crt -outform pem`

tormentist commented:
I can confirm my cert was already in PEM.  
Looks like this:
```
-----BEGIN CERTIFICATE-----
MIIFHTCCAw<redacted Base64 multiline data>
...
LGx0uH6uqc=
-----END CERTIFICATE-----

```
Running `openssl x509 -in old_pem_cert.crt -out tls_client.crt -outform pem` and comapring the files both visually or with `md5sum`, the files are identical, so it was indeed already in PEM.

For some reason, I couldn't get it to work with the tls_client.crt file in PEM format, kept getting this:  
`OpenSSLX509CertificateFactory$ParsingException: java.lang.RuntimeException: error:0c0000be:ASN.1 encoding routines:OPENSSL_internal:WRONG_TAG`
It ended up working after converting tls_client.crt to DER.

I cleared Cache and Data from App Info in-between each test too.  
Not sure how that'd be specific to my phone (Samsung Note 9), but I'll test on another phone just to be sure.

tormentist commented:
CER encoded tls_client.crt works fine on my SO's phone (Samsung Galaxy S20)
I started from scratch, uninstalled and reinstalled the apk on my phone and it works with CER encoded tls_client.crt now too.

So, everything is working as expected on both phones I can test.

Zero545 commented:
Tested the app-full-debug.apk  with a PEM-formatted tls_client.crt on my Huawei P30. Everything worked as expected.

carlos-sarmiento commented:
Tested the app as well with a Pixel 5 and it worked perfectly. Thanks for this @mircoboschi, my phone setup is going to be greatly simplified.

Now, fingers crossed to getting an equivalent feature on the iOS app

janost commented:
Working great on my Oneplus 9 Pro and Caddy's `client_auth`. Can't wait for this to be merged and released.

hllhll commented:
Awesome! Been waiting for this over 2 years! YT, Working!
Realme 7 5G
RMX2111
Android 11

evpaassen commented:
Tested this on a OnePlus 5T with Android 10 and I can confirm it works!

I noticed that the app requires the key to be an RSA key. If it is an EC key, the application crashes immediately, without any message. From the logs, I was able to extract the following exception that indicated that the key type was incorrect:

```
01-02 12:50:45.184 13269 13269 D AndroidRuntime: Shutting down VM
01-02 12:50:45.185 13269 13269 E AndroidRuntime: FATAL EXCEPTION: main
01-02 12:50:45.185 13269 13269 E AndroidRuntime: Process: io.homeassistant.companion.android.debug, PID: 13269
01-02 12:50:45.185 13269 13269 E AndroidRuntime: java.lang.RuntimeException: Unable to create application io.homeassistant.companion.android.HomeAssistantApplication: java.security.spec.InvalidKeySpecException: Unexpected key type
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6653)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.app.ActivityThread.access$1600(ActivityThread.java:232)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1953)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:107)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:214)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7697)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
01-02 12:50:45.185 13269 13269 E AndroidRuntime: Caused by: java.security.spec.InvalidKeySpecException: Unexpected key type
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at com.android.org.conscrypt.OpenSSLKey.getPrivateKey(OpenSSLKey.java:323)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at com.android.org.conscrypt.OpenSSLRSAKeyFactory.engineGeneratePrivate(OpenSSLRSAKeyFactory.java:74)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at java.security.KeyFactory.generatePrivate(KeyFactory.java:399)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at io.homeassistant.companion.android.common.data.MTLSHelper.tryImportKeys(MTLSHelper.kt:64)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at io.homeassistant.companion.android.common.data.MTLSHelper.init(MTLSHelper.kt:73)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at io.homeassistant.companion.android.HomeAssistantApplication.onCreate(HomeAssistantApplication.kt:40)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1204)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6648)
01-02 12:50:45.185 13269 13269 E AndroidRuntime:        ... 8 more
```

I think it would be useful to add this to the documentation (and even better would be to handle key type/format exceptions with a nice error message).

Anyway, it works for me now and thank you for your efforts making this work!

mircoboschi commented:
@evpaassen Thanks for your feedback.
I confirm right now only RSA algorithm is supported, i've update documentation.
Made some changes to my PR:
- Added successful import message
- Added file deletion after key import
- Added error message on import exception, to avoid crashes and improve user experience
- Resolved strings.xml conflict

Zero545 commented:
Is there still any activity on this pull request? I am really waiting this feature to be released.

mircoboschi commented:
Only waiting to be merged, on my side.

JBassett commented:
I was on vacation, sorry for the delay.  I'm not saying we need to do this for the initial PR of this feature but I think we want to move away from always loading a cert from a specific location and have a dialog that lets the user select the certificate once.  I'll try to get to reviewing this next free chance I get.

JBassett commented on common/src/main/java/io/homeassistant/companion/android/common/data/MTLSHelper.kt:
```diff
@@ -0,0 +1,145 @@
+    }
+    private fun getMTLSKeyManagerForOKHTTP(): X509KeyManager {
+        val mtlsKeyManager = object : X509KeyManager {
+            override fun getClientAliases(
+                p0: String?,
+                p1: Array<out Principal>?
+            ): Array<String> {
+                return emptyArray()
+            }
+
+            override fun chooseClientAlias(
+                p0: Array<out String>?,
+                p1: Array<out Principal>?,
+                p2: Socket?
+            ): String {
+                return ""
+            }
+
+            override fun getServerAliases(
+                p0: String?,
+                p1: Array<out Principal>?
+            ): Array<String> {
+                return arrayOf()
+            }
+
+            override fun chooseServerAlias(
+                p0: String?,
+                p1: Array<out Principal>?,
+                p2: Socket?
+            ): String {
+                return ""
+            }
```
If we aren't using these can we add some comments as to why?

mircoboschi replied:
These needs to be implemented because X509KeyManager is Abstract; i'll add a relevant comment.
For our purposes these functions will never be called, these are only placeholders.

JBassett replied:
Based on the [docs](https://developer.android.com/reference/javax/net/ssl/X509KeyManager) we should be using [X509ExtendedKeyManager](https://developer.android.com/reference/javax/net/ssl/X509ExtendedKeyManager) instead.

mircoboschi commented on common/src/main/java/io/homeassistant/companion/android/common/data/MTLSHelper.kt:
```diff
@@ -0,0 +1,145 @@
+        clientPublicKey = keyStore.getCertificate(keyentryname) ?: return
+    }
+
+    private fun tryImportKeys(context: Context) {
```
I suggest to leave both systems; with this implementation, it works even on Wear OS; don't know if is possible to do some file pickup there.
I agree to do this, maybe after first merge

JBassett replied:
Maybe.... Given the headache we are dealing with for external/internal URL on wear I'm not sure if we want to add this to the mess.

mircoboschi commented on wear/src/main/java/io/homeassistant/companion/android/HomeAssistantApplication.kt:
```diff
@@ -4,13 +4,16 @@ import android.app.Application
     override fun onCreate() {
         super.onCreate()
 
+        MTLSHelper().init(this)
+
```
This call is there for two reasons:
* it uses "this" context to call getExternalFilesDir and obtain base path to load files
* it actually tries to import keys if exists
If we use hilt (which i was not aware of) we could postpone import on first usage on WebClient or first OkHttp call.
How am i supposed to obtain base dir there? I'm not an Android expert, as you can see.

JBassett replied:
> * it uses "this" context to call getExternalFilesDir and obtain base path to load files

You can pass in the context via hilt


> * it actually tries to import keys if exists
>   If we use hilt (which i was not aware of) we could postpone import on first usage on WebClient or first OkHttp call.
>   How am i supposed to obtain base dir there? I'm not an Android expert, as you can see

This is why I was saying we should probably move to a file picker.  It could prompt for the keys after the app tries to authenticate and sees the need for them.

mircoboschi commented on common/src/main/java/io/homeassistant/companion/android/common/data/MTLSHelper.kt:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    fun init(context: Context) {
```
Got it; need to study hilt.

mircoboschi commented on common/src/main/java/io/homeassistant/companion/android/common/data/MTLSHelper.kt:
```diff
@@ -0,0 +1,145 @@
+    private val filename = "tls_client"
+    private val keyentryname = "client_mtls_cert"
+    companion object {
+        var clientPrivateKey: PrivateKey? = null
+        var clientPublicKey: Certificate? = null
+        var importMsg: String? = null
+        var importErrorMsg: String? = null
```
You're right for clientPrivateKey and clientPublicKey, but how can i retain importMsg and importErrorMsg if we do initial import on application bootup and need to display these later?

dshokouhi replied:
> You're right for clientPrivateKey and clientPublicKey, but how can i retain importMsg and importErrorMsg if we do initial import on application bootup and need to display these later?

do you need to retain these in between application restarts? If so then shared preferences may be what you need? Either stored in shared preferences or a database if the variables need to be persisted when the application is killed or device is restarted

JBassett replied:
If you use hilt the object will stick around for as long as the application is running.

mircoboschi commented on common/src/main/java/io/homeassistant/companion/android/common/data/MTLSHelper.kt:
```diff
@@ -0,0 +1,145 @@
+import javax.net.ssl.X509KeyManager
+import javax.net.ssl.X509TrustManager
+
+class MTLSHelper {
```
Ok i need to explore hilt

al-esc commented:
Finding this open PR is great news, looking forward to seeing it merged.

Thank you for all the effort folks! It has motivated me to try and contribute to this project.

justindthomas commented:
Tried to build from this PR, but seem to be running into conflicts with recent changes. I'm having trouble with the `wear:validateSigningRelease` task. It seems like that should be irrelevant since I'm not building for a watch, but this is the first time I've used Android Studio and Kotlin so I'm probably doing something wrong.

Will try again when the newer changes from `master` are merged (I can build that fine).

biscuit-316 commented:
I found this PR while searching for ways to improve Home Assistant security. I just tried the [app](https://github.com/home-assistant/android/actions/runs/1647188679) with the HAproxy package within PFSense. And it works great. I really hope this feature is added.

i'm quit new to github. As far as I can tell, it's been a while since the last changes were made. Can someone tell me what we are waiting for?

dshokouhi commented:
> i'm quit new to github. As far as I can tell, it's been a while since the last changes were made. Can someone tell me what we are waiting for?

The PR needs to be rebased with latest master changes and conflicts need to be resolved. The PR is also waiting for the requested changes mentioned above to be completed. Then it will need to be reviewed again.

mircoboschi commented:
I'm still waiting answers to questions in code suggestion section.
I'm on vacation now; i can rebase it next week. @dshokouhi can we chat on discord dev channel to clarify, on next week?

dshokouhi commented:
> I'm still waiting answers to questions in code suggestion section.

Yea I think we need to wait for @JBassett to get back to you on our your questions I tried my best to answer what I could



> I'm on vacation now; i can rebase it next week. @dshokouhi can we chat on discord dev channel to clarify, on next week?

sure, we are always on discord so feel free to join whenever. If I am not around I am sure someone will be able to help and offer advice :)

The biggest change to this PR will be adding hilt as we had migrated from dagger to hilt while this PR was submitted.  If you need any help with hilt I suggest looking at past PRs that involved using hilt as that should help guide you :)

magicbuffer commented:
Thanks @mircoboschi  for the code! I messed around that PR these last few days and got it working with the Android KeyChain, meaning there is no need to drop the client certificate and key on the device, users can install the PKCS#12 via the Android settings - with a strong password if they so wish. The app then prompt the user for a key via `KeyChain.choosePrivateKeyAlias`. As far as I can see in term of settings there is literally nothing to persist as the user is prompted every time the app performs the initial connection (which is good).

@mircoboschi  @homeassistant : Is that something you are interested in? Also was there a reason why that PR did not rely on the Android KeyChani in the first place?

kmccmk9 commented:
> Thanks @mircoboschi for the code! I messed around that PR these last few days and got it working with the Android KeyChain, meaning there is no need to drop the client certificate and key on the device, users can install the PKCS#12 via the Android settings - with a strong password if they so wish. The app then prompt the user for a key via `KeyChain.choosePrivateKeyAlias`. As far as I can see in term of settings there is literally nothing to persist as the user is prompted every time the app performs the initial connection (which is good).
> 
> PS: My changes are here https://github.com/magicbuffer/home-assistant-android-app
> 
> @mircoboschi @homeassistant : Is that something you are interested in? Also was there a reason why that PR did not rely on the Android KeyChain in the first place?

I've been testing your changes and I have one thing to report thus far. Every time you open the app it prompts to select a certificate. Also after selecting that certificate, it says "Unable to connect to Home Assistant". Pressing the retry button of course resolves this. I haven't tested notifications or sensors to see if they're affected by the certificate request. I will report back as I keep evaluating.

magicbuffer commented:
Thank you for taking the time to test my changes :+1: 

Regarding the "Unable to connect" message, that is race condition between all `HTTP` request going out when the app starts. As a quick fix I've been trying to delay them until a certificate is provided via the `WebView` but without too much success as you can see. For some reason it had a great success rate when I was testing on the emulator in debug mode but when I switched to my hw device in release mode I started to see these message a lot. Maybe the `HTTP` timeouts are relaxed in debug mode?! The cleanest way I can see without too much knowledge of the app would be to attempt to connect to the server before all the other services responsible for these requests even starts, get a key if client authentication is requested and then continue.

Second regarding the sensors, I wasn't aware HA comes with a background service, and what you describe makes a lot of sense. The key selected via the `KeyChain` method is bound to an `Activity`, right now the key is bound to the `WebView`, so as soon the `WebView` gets deleted the access to the key must be revoked by the `KeyChain`. I believe it should be possible to use another long lived `Activity` (Maybe the application's or the one from the service).

I have a bit of time in the upcoming days and will look into both of these issues.

magicbuffer commented:
So i had a quick look and regarding 1) I added some code to remember and load the key as soon as the app starts.  It seems to work fine in the emulator (I don't have access to a device right now). The main issue I have is that I cannot block the main thread (it is needed to receive a connection from the `KeyChain`) and I didn't find a clean and easy way to postpone all the http activity until a key is loaded (when TLS client authentication is on). Has someone an idea on how to achieve this?

Regarding 2) it turned out the application context was already used when accessing the KeyChain, so I will need to debug this further with actual sensors in the background. 
```java
        // prompt the user for a key
        val kcac = MyKeyChainCallbackAlias(activity.applicationContext)
```

@kmccmk9 If you have some time would you mind trying out https://github.com/magicbuffer/home-assistant-android-app/commit/2d63ae7311fafd73788b34ca2f6d549dd5bcf0a1 see if it helps regarding 1)?

Thanks :+1:

kmccmk9 commented:
> So i had a quick look and regarding 1) I added some code to remember and load the key as soon as the app starts. It seems to work fine in the emulator (I don't have access to a device right now). The main issue I have is that I cannot block the main thread (it is needed to receive a connection from the `KeyChain`) and I didn't find a clean and easy way to postpone all the http activity until a key is loaded (when TLS client authentication is on). Has someone an idea on how to achieve this?
> 
> Regarding 2) it turned out the application context was already used when accessing the KeyChain, so I will need to debug this further with actual sensors update in the background.
> 
> ```java
>         // prompt the user for a key
>         val kcac = MyKeyChainCallbackAlias(activity.applicationContext)
> ```
> 
> @kmccmk9 If you have some time would you mind trying out [magicbuffer@2d63ae7](https://github.com/magicbuffer/home-assistant-android-app/commit/2d63ae7311fafd73788b34ca2f6d549dd5bcf0a1) see if it helps regarding 1)?
> 
> Thanks 👍

Sorry for the late reply. i did test your changes and it does address the choosing certificate issue. With regards to number 2, how can I help?

mircoboschi commented:
Hello, sorry but i'm moving home; don't have much free time these days.
@magicbuffer no specific motivation beside avoiding multiple certificate requests and make it work even in background / wearOS situation.
With my code, beside first import, nothing is asked to user.

bryanyork commented:
@mircoboschi will you be abandoning this PR? Would love to see this brought into the official app, @magicbuffer would you step in for a new PR if this is the case?

kmccmk9 commented:
@magicbuffer After continued testing, everything seems to be working as expected including the sensor information.

magicbuffer commented:
Thank you @kmccmk9 for testing and reporting back the results :+1: I am glad to hear it also works for the sensors update, to be honest I wasn't expecting this. I'll take some time next week to polish the code and  merge the upstream changes so that it remains up-to-date.

@mircoboschi @bryanyork I apologize for hijacking this PR, it is not my intention to depreciate nor nullify @mircoboschi  work on this. That being said t is my personal opinion the proper way to implement this is by leveraging the Androiid KeyChain. As I am also a user of the app needing this feature, I do care about the overall security and the initial code did not provide the safeguards I am looking for (i.e. import of a key without password and the key material storage). My take on this would be integrate my changes into @mircoboschi repo and update this PR accordingly, also both @mircoboschi  and I could be working on tyding up the code. Ultimately that is a decision the HA maintainers must take :heart:

biscuit-316 commented:
@magicbuffer, I tried your version of the app.
The certificate selection was quite simple. Your version of the app works with RSA and Elliptic Curve certificates. Which is definitely a plus.

I was curious about how the app would work for sensor updates. especially since someone from my family uses the app only for the sensors and location tracking.
The updates of the sensors is working for the most part. I tested it with the app in the foreground, in the background. I removed the app from the recent list and restarted my phone. After all of that, the sensors were still coming in. All these tests were done last Saturday, but now 2 days later I saw that the state of charge was no longer updated. Opening the app solved the problem. 

Another issue I found was that if there is no certificate installed on the phone. The screen will turn completely white. There is no error message that the certificate is missing. The same applies if the certificate has been removed. The app hangs on the initializing screen. I was expecting an error message with instructions.

dshokouhi commented:
> I removed the app from the recent list

on some devices this is translated as a force stop and will stop the sensor worker from updating



> but now 2 days later I saw that the state of charge was no longer updated. Opening the app solved the problem.

sounds like the app does not have background permissions granted

biscuit-316 commented:
> > I removed the app from the recent list
> 
> on some devices this is translated as a force stop and will stop the sensor worker from updating

I was still receiving sensor updates after I removed the app from the recent list. 
 
> > but now 2 days later I saw that the state of charge was no longer updated. Opening the app solved the problem.
> 
> sounds like the app does not have background permissions granted

I thought so too. When I first opened the app, it asked for permission to run in the background. And I do believe I allowed it. Google told me that when an app wants to run in the background, all I need to do is turn off battery optimization. That was the only thing I found for my Nokia smartphone. Battery optimization was disabled for this Home Assistant debug app. I'm not quite sure why it didn’t work.

kmccmk9 commented:
> > > I removed the app from the recent list
> > 
> > 
> > on some devices this is translated as a force stop and will stop the sensor worker from updating
> 
> I was still receiving sensor updates after I removed the app from the recent list.
> 
> > > but now 2 days later I saw that the state of charge was no longer updated. Opening the app solved the problem.
> > 
> > 
> > sounds like the app does not have background permissions granted
> 
> I thought so too. When I first opened the app, it asked for permission to run in the background. And I do believe I allowed it. Google told me that when an app wants to run in the background, all I need to do is turn off battery optimization. That was the only thing I found for my Nokia smartphone. Battery optimization was disabled for this Home Assistant debug app. I'm not quite sure why it didn’t work.

I have not been able to replicate this issue yet with @magicbuffer code set.

BTW, @magicbuffer would it make more sense to open a new PR using your changes since they are working and satisfy the initial requirements of the numerous reported issues? From there, any additional changes could then be submitted. At this point it might be a good idea to do this more iteratively? Just a thought.

magicbuffer commented:
> BTW, @magicbuffer would it make more sense to open a new PR using your changes since they are working and satisfy the initial requirements of the numerous reported issues? From there, any additional changes could then be submitted. At this point it might be a good idea to do this more iteratively? Just a thought.

Hey sorry for the delay got carried away by work these last few days. Sure I think it is a good idea at this point. I'll get on to it and expect a new PR by the end of the week.

magicbuffer commented:
Openned #2526 wich contains this PR + the KeyChain modification as I did not heard back from @mircoboschi.

JBassett commented:
Closing since this hasn't been touched in over a month.

virtualdj commented:
> Client cert also works in Wear OS, but is tricky to upload cert files.

@mircoboschi Can you please suggest how to load (and unload) a cert file in a Wear OS device? How did you do it during testing?

mircoboschi commented:
> > Client cert also works in Wear OS, but is tricky to upload cert files.
> 
> @mircoboschi Can you please suggest how to load (and unload) a cert file in a Wear OS device? How did you do it during testing?

@virtualdj My implementation was able to import certificate from application folder; i was able to put files inside right emulator folder with adb if i recall correctly.

virtualdj commented:
@mircoboschi Yeah, I see that your implementation loaded the certificates from file in code, but I was interested in how you can load certificates on the watch using ADB for testing purposes. And to unload them, too.
I cannot find such documentation for Wear OS, so I asked you if you already found out.

Your code deleted the original certificates after they have been loaded, so I suppose there has to be a manual way using ADB, too...

odu14ick commented:
@mircoboschi @virtualdj 
Sorry to bring this old one up, but have either of you figured importing the ca certificate on wear os? Thanks!

virtualdj commented:
@odu14ick Yes, the functionality has already been implemented in Home Assistant for Wear OS, you just have to export your certificate (be aware of the encryption algorithm it uses because not all are supported by Wear OS) as a PFX file which the Home Assistant for Android on the smartphone will ask you when configuring/onboarding the watch.

I'm using it everyday and it works perfectly!

Note: on another thread the admins complained of not using Github issues for assistance, but the HA community forums instead.
